hkmc2/shared/src/main/scala/hkmc2/codegen/llir/Llir.scala

```diff
@@ -175,32 +100,150 @@ enum Node:
+      case Node.LetMethodCall(xs, cls, method, args, body) =>
+        doc"let ${xs.map(allocIfNew).mkString(",")} = ${allocIfNew(cls)}.${allocIfNew(method)}(${args.map(mkDocument).mkString(",")}) in # ${mkDocument(body)}"
+      case Node.LetCall(xs, func, args, body) => 
+        doc"let* (${xs.map(allocIfNew).mkString(",")}) = ${allocIfNew(func)}(${args.map(mkDocument).mkString(",")}) in # ${mkDocument(body)}"
+  def mkDocument(defn: Func)(using Raise, Scope): Document =
+    def docParams(params: Ls[Local]): Document =
+      params.map(allocIfNew).mkString("(", ",", ")")
+    given Conversion[String, Document] = raw
+    val docFirst = doc"def ${allocIfNew(defn.name)}${docParams(defn.params)} ="
+    val docBody = mkDocument(defn.body)
+    doc"$docFirst #{  # $docBody #} "
+  def mkDocument(cls: ClassInfo)(using Raise, Scope): Document =
+    given Conversion[String, Document] = raw
+    val ext = if cls.parents.isEmpty then "" else " extends " + cls.parents.map(allocIfNew).mkString(", ")
+    val docFirst = doc"class ${allocIfNew(cls.symbol)}(${cls.fields.map(_.nme).mkString(",")})$ext {"
+    val docMethods = cls.methods.map { (_, func) => mkDocument(func) }.toList.mkDocument(doc" # ")
+    val docLast = doc"}"
+    doc"$docFirst #{  # $docMethods #}  # $docLast"
+  def mkDocument(prog: Program, hide: Str => Bool = defaultHidden)(using Raise, State): Document =
+    given Scope = Scope.empty
     given Conversion[String, Document] = raw
-    this match
-    case Result(res) => (res |> showArguments)
-    case Jump(jp, args) =>
-      doc"jump ${docSymWithUid(jp)}(${args |> showArguments})"
-    case Case(x, cases, default) =>
-      val docFirst = doc"case ${x.toString} of"
-      val docCases = cases.map {
-        case (pat, node) => doc"${pat.toString} => #{  # ${node.toDocument} #} "
-      }.mkDocument(doc" # ")
-      default match
-        case N => doc"$docFirst #{  # $docCases #} "
-        case S(dc) =>
-          val docDeft = doc"_ => #{  # ${dc.toDocument} #} "
-          doc"$docFirst #{  # $docCases # $docDeft #} "
-    case Panic(msg) =>
-      doc"panic ${s"\"$msg\""}"
-    case LetExpr(x, expr, body) => 
-      doc"let ${docSymWithUid(x)} = ${expr.toString} in # ${body.toDocument}"
-    case LetMethodCall(xs, cls, method, args, body) =>
-      doc"let ${xs.map(docSymWithUid).mkString(",")} = ${cls.nme}.${docSymWithUid(method)}(${args.map(_.toString).mkString(",")}) in # ${body.toDocument}"
-    case LetCall(xs, func, args, body) => 
-      doc"let* (${xs.map(docSymWithUid).mkString(",")}) = ${func.nme}(${args.map(_.toString).mkString(",")}) in # ${body.toDocument}"
+    val docClasses = prog.classes.filterNot(c => hide(c.symbol.nme)).map(mkDocument).toList.mkDocument(doc" # ")
+    val docDefs = prog.defs.map(mkDocument).toList.mkDocument(doc" # ")
+    val docMain = doc"entry = ${allocIfNew(prog.entry)}"
+    doc" #{ $docClasses\n$docDefs\n$docMain #} "
 
+object LlirDebugPrinter:
```

Comment:
There's a lot of duplication between these two. Isn't most of the logic identical? In such case, it's better to put most of the logic in a common class and selectively override or implement some methods in the derived objects.